<!DOCTYPE html>

<html lang="en">
<head>
<!-- Basic Page Needs
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
<meta charset="utf-8"/>
<title>GSForge.operations.normalizations — GSForge 0.0.1 documentation</title>
<meta content="Short description for html meta description." name="description"/>
<meta content="Tyler Biggs" name="author"/>
<!-- Mobile Specific Metas
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
<meta content="width=device-width, initial-scale=1" name="viewport"/>
<!-- FONT
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
<script src="https://ajax.googleapis.com/ajax/libs/webfont/1.6.16/webfont.js"></script>
<script>
    WebFont.load({
      google: {
        families: ['Source Sans Pro']
      }
    });
  </script>
<!-- CSS
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
<link href="../../../_static/css/main.css" rel="stylesheet"/>
<link href="../../../_static/nbsite.css" rel="stylesheet"/>
<!-- Scripts
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
<script src="//ajax.googleapis.com/ajax/libs/jquery/2.1.1/jquery.min.js"></script>
<script async="" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML" type="text/javascript">
</script>
<script src="../../../_static/js/main.js"></script>
<script src="../../../_static/nbsite.js"></script>
<script src="../../../_static/require.js"></script>
<!-- Favicon
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
<link href="../../../_static/images/favicon.ico" rel="icon" type="image/png"/>
<!-- Canonical
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
<link href="/_modules/GSForge/operations/normalizations.html" rel="canonical">
</link></head>
<body class="">
<header class="navigation">
<div class="wrapper">
<a class="logo" href="../../../index.html">
<img alt="Logo" src="../../../_static/images/logo.png"/>
</a>
<a class="navigation-menu logo-text" href="../../../index.html">GSForge</a>
<a class="navigation-menu-button" href="javascript:void(0)" id="js-mobile-menu">Menu</a>
<nav>
<ul class="navigation-menu show" id="js-navigation-menu">
<li class="nav-link"><a href="../../../Welcome.html">Welcome</a></li>
<li class="nav-link"><a href="../../../user_guide/index.html">User Guide</a></li>
<li class="nav-link"><a href="../../../Reference_Manual/index.html">API</a></li>
<li class="nav-link"><a href="../../../Development.html">Developer Guide</a></li>
<li class="nav-link"><a href="../../../About.html">About</a></li>
<li class="nav-link"><div style="display:inline-block;vertical-align: middle;"><div class="search-bar"><form action="../../../search.html" method="get" role="search">
<input name="q" placeholder="Search" type="search"/>
<button type="submit">
<img alt="Search Icon" src="https://raw.githubusercontent.com/thoughtbot/refills/master/source/images/search-icon-black.png"/>
</button>
</form>
</div></div></li>
</ul>
</nav>
</div>
</header>
<!-- MAIN BODY OF DOCS –––––––––––––––––– -->
<div class="docs section">
<div id="hacketyhackhack"> <!-- style="width:20%;margin-right: 100px;"> --> <!--style="display: none;"> style="display:none;"> -->
<div class="toc" style="width:15%; margin-right:20px;">
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../user_guide/index.html">User Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Reference_Manual/index.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../About.html">About</a></li>
</ul>
</div>
</div>
<div class="content"> <!-- style="max-width:80%;margin-left:auto;margin-right: auto;">-->
<h1>Source code for GSForge.operations.normalizations</h1><div class="highlight"><pre>
<span></span><span class="sd">"""</span>
<span class="sd">Normalization functions inherit from the ``OperationInterface`` class.</span>
<span class="sd">This means that they can all be called upon an ``AnnotatedGEM`` or a ``GeneSetCollection``.</span>

<span class="sd">These (classes) functions have static methods that implement the transform on a ``numpy``</span>
<span class="sd">or ``xarray`` source.</span>
<span class="sd">"""</span>

<span class="kn">import</span> <span class="nn">param</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">..models</span> <span class="kn">import</span> <span class="n">OperationInterface</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">"ReadsPerKilobaseMillion"</span><span class="p">,</span>
    <span class="s2">"UpperQuartile"</span><span class="p">,</span>
<span class="p">]</span>


<div class="viewcode-block" id="UpperQuartile"><a class="viewcode-back" href="../../../Reference_Manual/GSForge.operations.html#GSForge.operations.normalizations.UpperQuartile">[docs]</a><span class="k">class</span> <span class="nc">UpperQuartile</span><span class="p">(</span><span class="n">OperationInterface</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    Under this normalization method, after removing genes having zero read counts for all samples,</span>
<span class="sd">    the remaining gene counts are divided by the upper quartile of counts different from zero in</span>
<span class="sd">    the computation of the normalization factors associated with their sample and multiplied by</span>
<span class="sd">    the mean upper quartile across all samples of the dataset. [method_compare]_</span>

<span class="sd">    Original R code.  ::</span>

<span class="sd">        uq&lt;-function(X){</span>

<span class="sd">          #excluding zero counts in each sample</span>
<span class="sd">          UQ&lt;-function(y){</span>
<span class="sd">            quantile(y, 0.75)</span>
<span class="sd">          }</span>
<span class="sd">          X&lt;-X+0.1</span>
<span class="sd">          upperQ&lt;-apply(X,2,UQ)</span>
<span class="sd">          f.uq&lt;-upperQ/mean(upperQ)</span>
<span class="sd">          upq.res&lt;-scale(X,center=FALSE,scale=f.uq)</span>
<span class="sd">          return(upq.res)</span>
<span class="sd">        }</span>

<span class="sd">    .. [method_compare] `A comparison of per sample global scaling and per gene normalization methods for differential expression analysis of RNA-seq data &lt;https://doi.org/10.1371/journal.pone.0176185&gt;`_</span>

<span class="sd">    """</span>

    <span class="c1"># For clarity of testing I write normalization methods as their own, self-contained functions.</span>
    <span class="c1"># I often create a numpy and xarray version.</span>
    <span class="c1"># @staticmethods decorated functions do not depend on the state of the class, and may be called</span>
    <span class="c1"># without instantiating the class.</span>
<div class="viewcode-block" id="UpperQuartile.np_upper_quartile"><a class="viewcode-back" href="../../../Reference_Manual/GSForge.operations.html#GSForge.operations.normalizations.UpperQuartile.np_upper_quartile">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">np_upper_quartile</span><span class="p">(</span><span class="n">counts</span><span class="p">):</span>
        <span class="sd">"""</span>
<span class="sd">        Perform the upper quartile normalization.</span>

<span class="sd">        :param counts: A numpy array containing the raw count values. The shape is assumed to be</span>
<span class="sd">            (samples by genes). Zero counts are expected to be present as zeros.</span>

<span class="sd">        :returns: The upper quartile normalized count matrix.</span>
<span class="sd">        """</span>
        <span class="n">adjusted_counts</span> <span class="o">=</span> <span class="n">counts</span> <span class="o">+</span> <span class="mf">0.1</span>
        <span class="n">per_sample_quantile</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">adjusted_counts</span><span class="p">,</span> <span class="n">q</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">per_sample_quantile_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">per_sample_quantile</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">adjusted_counts</span> <span class="o">/</span> <span class="p">(</span><span class="n">per_sample_quantile</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span> <span class="o">/</span> <span class="n">per_sample_quantile_mean</span><span class="p">)</span></div>

<div class="viewcode-block" id="UpperQuartile.xr_upper_quartile"><a class="viewcode-back" href="../../../Reference_Manual/GSForge.operations.html#GSForge.operations.normalizations.UpperQuartile.xr_upper_quartile">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">xr_upper_quartile</span><span class="p">(</span><span class="n">counts</span><span class="p">):</span>
        <span class="sd">"""</span>
<span class="sd">        Perform the upper quartile normalization.</span>

<span class="sd">        :param counts: An ``xarray.DataArray`` containing the raw count values. The shape is</span>
<span class="sd">            assumed to be (samples by genes). Zero counts are expected to be present as zeros.</span>

<span class="sd">        :returns: The upper quartile normalized count matrix.</span>
<span class="sd">        """</span>
        <span class="n">adjusted_counts</span> <span class="o">=</span> <span class="n">counts</span> <span class="o">+</span> <span class="mf">0.1</span>
        <span class="n">per_sample_quantile</span> <span class="o">=</span> <span class="n">adjusted_counts</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="s2">"Gene"</span><span class="p">)</span>
        <span class="n">per_sample_quantile_mean</span> <span class="o">=</span> <span class="n">per_sample_quantile</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">adjusted_counts</span> <span class="o">/</span> <span class="p">(</span><span class="n">per_sample_quantile</span> <span class="o">/</span> <span class="n">per_sample_quantile_mean</span><span class="p">))</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">"quantile"</span><span class="p">)</span></div>

<div class="viewcode-block" id="UpperQuartile.process"><a class="viewcode-back" href="../../../Reference_Manual/GSForge.operations.html#GSForge.operations.normalizations.UpperQuartile.process">[docs]</a>    <span class="k">def</span> <span class="nf">process</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">"""</span>
<span class="sd">        Perform the upper quartile normalization.</span>

<span class="sd">        :return: The upper quartile normalized count matrix.</span>
<span class="sd">        """</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">x_count_data</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">xr_upper_quartile</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="ReadsPerKilobaseMillion"><a class="viewcode-back" href="../../../Reference_Manual/GSForge.operations.html#GSForge.operations.normalizations.ReadsPerKilobaseMillion">[docs]</a><span class="k">class</span> <span class="nc">ReadsPerKilobaseMillion</span><span class="p">(</span><span class="n">OperationInterface</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    RPKM or FPKM -- Reads or Fragments per per Kilobase Million.</span>

<span class="sd">    These methods attempt to compensate for sequencing depth and gene length.</span>
<span class="sd">    The utility of this method is disputed in the literature [cite me].</span>
<span class="sd">    """</span>
    <span class="n">length_variable</span> <span class="o">=</span> <span class="n">param</span><span class="o">.</span><span class="n">String</span><span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="s2">"lengths"</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">xr_reads_per_kilobase_million</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">lengths</span><span class="p">,</span> <span class="n">sample_dim</span><span class="o">=</span><span class="s2">"Sample"</span><span class="p">):</span>
        <span class="n">scaling_factor</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="n">sample_dim</span><span class="p">)</span> <span class="o">/</span> <span class="mf">1e6</span>
        <span class="n">reads_per_million</span> <span class="o">=</span> <span class="n">counts</span> <span class="o">/</span> <span class="n">scaling_factor</span>
        <span class="n">normalized_counts</span> <span class="o">=</span> <span class="n">reads_per_million</span> <span class="o">/</span> <span class="n">lengths</span>
        <span class="k">return</span> <span class="n">normalized_counts</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">np_reads_per_kilobase_million</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">lengths</span><span class="p">):</span>
        <span class="n">scaling_factor</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">/</span> <span class="mf">1e6</span>
        <span class="n">reads_per_million</span> <span class="o">=</span> <span class="n">counts</span> <span class="o">/</span> <span class="n">scaling_factor</span>
        <span class="n">normalized_counts</span> <span class="o">=</span> <span class="n">reads_per_million</span> <span class="o">/</span> <span class="n">lengths</span>
        <span class="k">return</span> <span class="n">normalized_counts</span>

<div class="viewcode-block" id="ReadsPerKilobaseMillion.process"><a class="viewcode-back" href="../../../Reference_Manual/GSForge.operations.html#GSForge.operations.normalizations.ReadsPerKilobaseMillion.process">[docs]</a>    <span class="k">def</span> <span class="nf">process</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">x_count_data</span>
        <span class="n">lengths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gem</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">length_variable</span><span class="p">]</span><span class="o">.</span><span class="n">sel</span><span class="p">({</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_index_name</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_index</span><span class="p">()})</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">xr_reads_per_kilobase_million</span><span class="p">(</span><span class="n">counts</span><span class="o">=</span><span class="n">counts</span><span class="p">,</span> <span class="n">lengths</span><span class="o">=</span><span class="n">lengths</span><span class="p">,</span> <span class="n">sample_dim</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_index_name</span><span class="p">)</span></div></div>
</pre></div>
</div>
</div>
<!-- END MAIN BODY OF DOCS ––––––––––––– -->
<footer class="footer">
<div class="footer-links">
<ul>
<li><span class="footer-title">Links</span></li>
<li><a href="../../../Welcome.html">Welcome</a></li>
<li><a href="../../../user_guide/index.html">User Guide</a></li>
<li><a href="../../../Reference_Manual/index.html">API</a></li>
<li><a href="../../../Development.html">Developer Guide</a></li>
<li><a href="../../../About.html">About</a></li>
</ul>
<ul>
<li><span class="footer-title">Join Us</span></li>
<li><a href="https://github.com/SystemsGenetics/GSForge">Github</a></li>
</ul>
<ul class="copyright">
<li>2019 Tyler Biggs</li>
</ul>
</div>
</footer>
<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-61554933-1', 'auto');
  ga('send', 'pageview');

</script>
<!-- MathJax Config -->
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}
});
</script>
<!-- End Document
  –––––––––––––––––––––––––––––––––––––––––––––––––– -->
</body>
</html>